server/src/main/java/org/opensearch/index/store/remote/filecache/FullFileCachedIndexInput.java

```diff
@@ -88,21 +87,37 @@ public void close() throws IOException {
             if (isClone) {
                 cache.decRef(filePath);
             }
-            clones.forEach(indexInput -> {
-                try {
-                    indexInput.close();
-                } catch (Exception e) {
-                    logger.trace("Exception while closing clone - {}", e.getMessage());
-                }
-            });
             try {
                 luceneIndexInput.close();
             } catch (AlreadyClosedException e) {
                 logger.trace("FullFileCachedIndexInput already closed");
             }
             luceneIndexInput = null;
-            clones.clear();
             closed = true;
         }
     }
+
+    private static class IndexInputHolder implements Runnable {
+        private final IndexInput indexInput;
+        private final FileCache cache;
+        private final boolean isClone;
+        private final Path path;
+
+        IndexInputHolder(IndexInput indexInput, boolean isClone, FileCache cache, Path path) {
+            this.indexInput = indexInput;
+            this.isClone = isClone;
+            this.cache = cache;
+            this.path = path;
+        }
+
+        @Override
+        public void run() {
+            try {
+                indexInput.close();
```

Comment:
would this automatically do `decRef`  ?

Reply:
Oh now i get it . It is `underlyingIndexInput` only and not `FullFileCachedIndexInput`.